### DISCUSSION TITLE
Save as `png`

### DISCUSSION THREAD

**Discussion by t-bltg** [FedeClaudi/Term.jl#121] [Ideas]
It would be awesome (but - I guess - not easy to implement) to be able to save `Renderable`s as `png` images.

For `UnicodePlots`, it was added in  https://github.com/JuliaPlots/UnicodePlots.jl/pull/239.

And for `Plots.jl` we save each `Plot` as png, and insert it into a bigger image for simple non-nested grid layouts: https://github.com/JuliaPlots/Plots.jl/blob/ec77b9728bd09bfe8e0a511d32268373d6a1fc83/src/backends/unicodeplots.jl#L257

**Comment by FedeClaudi** [comment]
It would be great. 
As you say in the PR linked it would also make documentation easier. 

It should be possible. `rich` in python even exports to SVG and HTML. 
How/where is this done in UnicodePlots? Would it be hard to adapt? Is it worth making a separate tool taking terminal output and converting it to image?

**Comment by t-bltg** [comment]
For `UnicodePlots`, I use `FreeTypeAbstraction` to render each plot row into a `Matrix` of `RGBA` colors (https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L1015). Each character is thus rendered using freetype and drawn into this matrix.

It's satisfactory for now, and this is how I generate the documentation images such as https://raw.githubusercontent.com/JuliaPlots/UnicodePlots.jl/unicodeplots-docs/2.10/surfaceplot1.png in [README.md](https://github.com/JuliaPlots/UnicodePlots.jl#readme) (expect for the [front banner](https://raw.githubusercontent.com/JuliaPlots/UnicodePlots.jl/unicodeplots-docs/2.10/banner.jpg) for which I used `Term` and thus had to take a screenshot - hence this discussion). One can increase the size of the final image (resolution) using the [pixelsize](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L867) argument which controls the size of each terminal character.

In order to do that with Term we'd have to [estimate the final image size](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L930), [create it](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L1003), [split the renderable lines into `chars` and `colors`](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L965) for the foreground & background, [render each row into this image](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L1015), and [save it](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L1063) using `FileIO`.

I hack [`_show`](https://github.com/JuliaPlots/UnicodePlots.jl/blob/df510da8407c52c91cd0ddd2e83f4498f7e34839/src/plot.jl#L965) in order to split the chars and colors, so I guess this is too `UnicodePlots` specific to be split out into a package for usage in `Term`, however `FreeTypeAbstraction` is a small, low deps and contained dependency that can be added to `Term`.

**Comment by FedeClaudi** [comment]
Got it. It sounds interesting, but to be honest my bandwidth is a bit limited at the moment (finishing the PhD and what not). 
I'll be on holiday from the end of the week and will be back with the batteries a bit recharged. Ping me in a couple week if I haven't got back to you on this, just in case I forgot.

**Comment by t-bltg** [comment]
> finishing the PhD

Been there too, priorities ;)

Yes, this is my 2 cents on this so anyone interested can help, I'll see if I have time to propose something but I'm afraid this is too time consuming for me to implement.

**Reply by FedeClaudi** [reply]
yeah that's fair. Let's table for it now, I'll come back to it when I have more time
